### ISSUE TITLE
require Primitive.get_original_require(__FILE__) not working for Concurrent::AtomicReference patch

### ISSUE THREAD

**Issue by eregon** [oracle/truffleruby#2605]
From https://github.com/sinatra/sinatra/pull/1751#issuecomment-1050535025

I can reproduce the issue locally with sinatra/sinatra@0309bbb9b28044bdde9f96ecae811f6534c7c851:
```
cd sinatra
git pull # see commit I used above
bundle config --local path vendor/bundle       
bundle install
bundle exec rake test TEST=test/rabl_test.rb   
```
which gives:
```
[To redirect Truffle log output to a file use one of the following options:
* '--log.file=<path>' if the option is passed using a guest language launcher.
* '-Dpolyglot.log.file=<path>' if the option is passed using the host Java launcher.
* Configure logging using the polyglot embedding API.]
[ruby] WARNING: Encoding.find('locale') is US-ASCII, this often indicates that the system locale is not set properly. Set LANG=C and LC_ALL=C to suppress this warning (but some things might break).
[:in, "/home/eregon/.rubies/truffleruby-dev/lib/patches/concurrent/atomic/atomic_reference.rb", :original, "/home/eregon/code/sinatra/vendor/bundle/truffleruby/3.0.2.13/gems/concurrent-ruby-1.1.9/lib/concurrent-ruby/concurrent/atomic/atomic_reference.rb"]
/home/eregon/.rubies/truffleruby-dev/lib/patches/concurrent/atomic/atomic_reference.rb:10: warning: loading in progress, circular require considered harmful - /home/eregon/code/sinatra/vendor/bundle/truffleruby/3.0.2.13/gems/concurrent-ruby-1.1.9/lib/concurrent-ruby/concurrent/atomic/atomic_reference.rb
/home/eregon/code/sinatra/vendor/bundle/truffleruby/3.0.2.13/gems/concurrent-ruby-1.1.9/lib/concurrent-ruby/concurrent/configuration.rb:89:in `const_missing': uninitialized constant Concurrent::AtomicReference (NameError)
	from /home/eregon/code/sinatra/vendor/bundle/truffleruby/3.0.2.13/gems/concurrent-ruby-1.1.9/lib/concurrent-ruby/concurrent/configuration.rb:89:in `<module:Concurrent>'
	from /home/eregon/code/sinatra/vendor/bundle/truffleruby/3.0.2.13/gems/concurrent-ruby-1.1.9/lib/concurrent-ruby/concurrent/configuration.rb:11:in `<top (required)>'
	from <internal:core> core/kernel.rb:234:in `gem_original_require'
	from /home/eregon/code/sinatra/vendor/bundle/truffleruby/3.0.2.13/gems/concurrent-ruby-1.1.9/lib/concurrent-ruby/concurrent.rb:4:in `<top (required)>'
	from <internal:core> core/kernel.rb:234:in `gem_original_require'
	from /home/eregon/code/sinatra/vendor/bundle/truffleruby/3.0.2.13/gems/activesupport-6.1.4.6/lib/active_support/logger_thread_safe_level.rb:5:in `<top (required)>'
	from <internal:core> core/kernel.rb:234:in `gem_original_require'
	from /home/eregon/code/sinatra/vendor/bundle/truffleruby/3.0.2.13/gems/activesupport-6.1.4.6/lib/active_support/logger_silence.rb:5:in `<top (required)>'
	from <internal:core> core/kernel.rb:234:in `gem_original_require'
	from /home/eregon/code/sinatra/vendor/bundle/truffleruby/3.0.2.13/gems/activesupport-6.1.4.6/lib/active_support/logger.rb:3:in `<top (required)>'
	from <internal:core> core/kernel.rb:234:in `gem_original_require'
	from /home/eregon/code/sinatra/vendor/bundle/truffleruby/3.0.2.13/gems/activesupport-6.1.4.6/lib/active_support.rb:29:in `<top (required)>'
	from <internal:core> core/kernel.rb:234:in `gem_original_require'
	from /home/eregon/code/sinatra/vendor/bundle/truffleruby/3.0.2.13/gems/rabl-0.15.0/lib/rabl.rb:1:in `<top (required)>'
	from <internal:core> core/kernel.rb:234:in `gem_original_require'
	from /home/eregon/code/sinatra/test/rabl_test.rb:4:in `<top (required)>'
	from <internal:core> core/kernel.rb:234:in `gem_original_require'
	from /home/eregon/code/sinatra/vendor/bundle/truffleruby/3.0.2.13/gems/rake-13.0.6/lib/rake/rake_test_loader.rb:21:in `block in <main>'
	from /home/eregon/code/sinatra/vendor/bundle/truffleruby/3.0.2.13/gems/rake-13.0.6/lib/rake/rake_test_loader.rb:6:in `select'
	from /home/eregon/code/sinatra/vendor/bundle/truffleruby/3.0.2.13/gems/rake-13.0.6/lib/rake/rake_test_loader.rb:6:in `<main>'
```

I added `p [:in, __FILE__, :original, Primitive.get_original_require(__FILE__)]` in `lib/patches/concurrent/atomic/atomic_reference.rb`.
So it's weird because that shows:
```
[:in, "/home/eregon/.rubies/truffleruby-dev/lib/patches/concurrent/atomic/atomic_reference.rb", :original, "/home/eregon/code/sinatra/vendor/bundle/truffleruby/3.0.2.13/gems/concurrent-ruby-1.1.9/lib/concurrent-ruby/concurrent/atomic/atomic_reference.rb"]
```
but then the later `require Primitive.get_original_require(__FILE__)` seems to load the patch file again somehow, even though we give it an absolute path??
Adding a debug print in `vendor/bundle/truffleruby/3.0.2.13/gems/concurrent-ruby-1.1.9/lib/concurrent-ruby/concurrent/atomic/atomic_reference.rb` it's never showned so never loaded.

@bjfish Could you investigate?

**Comment by bjfish** [comment]
In the following `relative_feature` method, the absolute path: `/Users/bfish/Documents/projects/sinatra/vendor/bundle/truffleruby/3.0.2.13/gems/concurrent-ruby-1.1.9/lib/concurrent-ruby/concurrent/atomic/atomic_reference.rb`
is getting transformed to `vendor/bundle/truffleruby/3.0.2.13/gems/concurrent-ruby-1.1.9/lib/concurrent-ruby/concurrent/atomic/atomic_reference` instead of the desired `concurrent/atomic/atomic_reference` because `/Users/bfish/Documents/projects/sinatra` appears on the load path before `/Users/bfish/Documents/projects/sinatra/vendor/bundle/truffleruby/3.0.2.13/gems/concurrent-ruby-1.1.9/lib/concurrent-ruby`.

https://github.com/oracle/truffleruby/blob/8d9858227d2fcd38304e760f5112cdc5b3f2f628/src/main/ruby/truffleruby/core/truffle/feature_loader.rb#L221

One idea for this case would be to find all matches in the load path and return the one with the shortest resulting relative path.

**Comment by eregon** [comment]
Good find, could you try that?
`relative_feature` is heuristics and unfortunately it seems hard to make it fully reliable.

I'm not sure how we end up not loading the original file though, since we do find the patch file initially and immediately do `patchFiles.put(relativeFeature, true);`.

**Comment by bjfish** [comment]
@eregon Yes, I'll try this. I think it's returning before loading since it thinks there's a circular require:
`lock.isHeldByCurrentThread()lock.isHeldByCurrentThread()` is `true` but since the `relativeFeature` is wrong `isPatched` is `false`
https://github.com/oracle/truffleruby/blob/8d9858227d2fcd38304e760f5112cdc5b3f2f628/src/main/java/org/truffleruby/language/loader/RequireNode.java#L154-L168

**Comment by eregon** [comment]
I realized, `/Users/bfish/Documents/projects/sinatra` shouldn't be in $LOAD_PATH, `/Users/bfish/Documents/projects/sinatra/lib` should instead and that makes sense.
And that would avoid the issue altogether.

**Comment by bjfish** [comment]
@eregon This is added here: https://github.com/sinatra/sinatra/blob/ae70708d21db79de7d8ef8195d356c2fa76a2d3a/Rakefile#L29-L40

**Comment by eregon** [comment]
Can you try to remove it and see if that works? i.e. is it actually needed there?

**Comment by bjfish** [comment]
@eregon The tests passed without this. PR submitted here: https://github.com/sinatra/sinatra/pull/1765

**Comment by bjfish** [comment]
This is fixed at https://github.com/oracle/truffleruby/commit/8f5d1fcc6c9b7c800970497034aefd074de951e8.
